MoreLinq/TrySingle.cs

```diff
@@ -0,0 +1,171 @@
+        /// cardinality. This enables the caller to differentiate between the
+        /// case where a sequence contains zero elements and many elements.
+        /// </summary>
+        /// <param name="source">
+        /// The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has zero
+        /// elements.</param>
+        /// <param name="one">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has a
+        /// single element only.</param>
+        /// <param name="many">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has two or
+        /// more elements.</param>
+        /// <param name="resultSelector">
+        /// A function that is provided with the cardinality, and if the
+        /// sequence has just one element, the value of that element. Then
+        /// transforms the result to an instance of TResult.</param>
+        /// <typeparam name="T">
+        /// The type of the elements of the sequence.</typeparam>
+        /// <typeparam name="TCardinality">
+        /// The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">
+        /// The result type of the <paramref name="resultSelector"/> function.
+        /// </typeparam>
+        /// <returns>
+        /// The value returned by the <paramref name="resultSelector"/>.
+        /// </returns>
+        /// <remarks>
+        /// This operator uses immediate execution, but never consumes more
+        /// than two elements from the sequence. When the source sequence is an
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
+            TCardinality zero, TCardinality one, TCardinality many,
+            Func<TCardinality, T, TResult> resultSelector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            switch (source)
+            {
+                case IReadOnlyCollection<T> readOnlyCollection:
+                    return TrySingleForReadOnlyCollection(readOnlyCollection);
+                case ICollection<T> collection:
+                    return TrySingleForCollection(collection);
+                default:
+                    return TrySingleForEnumerable(source);
+            }
+
+            TResult TrySingleForReadOnlyCollection(IReadOnlyCollection<T> theCollection)
```

Comment:
This is largely redundant with `TrySingleForCollection`. I think you could just fold all the local functions into a single 3-level switch that it really is. Like this?

```c#
            switch (source.TryGetCollectionCount())
            {
                case int count:
                {
                    switch (count)
                    {
                        case 0:
                            return resultSelector(zero, default);
                        case 1:
                            T item;
                            switch (source)
                            {
                                case IReadOnlyList<T> list: item = list[0]; break;
                                case IList<T> list: item = list[0]; break;
                                default: item = source.First(); break;
                            }
                            return resultSelector(one, item);
                        default:
                            return resultSelector(many, default);
                    }
                }
                default:
                {
                    using (var e = source.GetEnumerator())
                    {
                        if (!e.MoveNext())
                            return resultSelector(zero, default);
                        var current = e.Current;
                        return !e.MoveNext() ? resultSelector(one, current)
                                             : resultSelector(many, default);
                    }
                }
            }
```